File: python/day03.py
#!/usr/bin/env python3
'''Advent of Code 2024 Day 3 solution'''
import argparse
import re

# Surprisingly efficient to use regexps with ^, the runtime is about 20ms on the full problem.
MUL = re.compile(r"^mul\(([0-9]+),([0-9]+)\)")
DO = re.compile(r"^do\(\)")
DONT = re.compile(r"^don't\(\)")

def runpart1(data):
    '''Run part one'''
    total = 0
    for i in range(len(data)):
        r = MUL.match(data[i-1:])
        if r:
            total += int(r.group(1)) * int(r.group(2))

    return total

def runpart2(data):
    '''Run part two'''
    enabled = True

    total = 0
    for i in range(len(data)):
        r = MUL.match(data[i-1:])
        if r and enabled:
            total += int(r.group(1)) * int(r.group(2))
        elif DO.match(data[i-1:]):
            enabled = True
        elif DONT.match(data[i-1:]):
            enabled = False

    return total

def main():
    '''Main'''
    parser = argparse.ArgumentParser()
    parser.add_argument("filename", nargs='?', default="inputs/day03.txt")
    args = parser.parse_args()

    # Read inputa file
    with open(args.filename, 'r', encoding="utf8") as file:
        data = file.read()

    # Run the solution
    part1 = runpart1(data)
    part2 = runpart2(data)
    return (part1, part2)

if __name__ == '__main__':
    print(main())

File: golang/day03/day03.go
package main

import (
	"flag"
	"fmt"
	"log"
	"os"
	"strings"
)

func readData(inputFile string) string {
	data, err := os.ReadFile(inputFile)
	if err != nil {
		log.Fatal(err)
		os.Exit(1)
	}

	return (string(data))
}

func runPart1(data string) int {
	var a, b int
	total := 0

	for i := range data {
		_, err := fmt.Sscanf(data[i:], "mul(%d,%d)", &a, &b)
		if err == nil {
			total += a * b
		}
	}

	return total
}

func runPart2(data string) int {
	var a, b int
	total := 0
	enabled := true

	for i := range data {
		if strings.HasPrefix(data[i:], "do()") {
			enabled = true
		} else if strings.HasPrefix(data[i:], "don't()") {
			enabled = false
		} else if enabled {
			_, err := fmt.Sscanf(data[i:], "mul(%d,%d)", &a, &b)
			if err == nil {
				total += a * b
			}
		}
	}

	return total
}

func main() {
	var inputFile = flag.String("input", "inputs/day03.txt", "Problem input file")
	flag.Parse()
	data := readData(*inputFile)
	part1 := runPart1(data)
	part2 := runPart2(data)
	fmt.Printf("%d, %d\n", part1, part2)
}
